test/e2e_node/container_restart_test.go

```diff
@@ -0,0 +1,121 @@
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	"k8s.io/kubernetes/test/e2e/feature"
+	"k8s.io/kubernetes/test/e2e/framework"
+	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
+	admissionapi "k8s.io/pod-security-admission/api"
+)
+
+const containerName = "restarts"
+
+var _ = SIGDescribe("Container Restart", feature.CriProxy, framework.WithSerial(), func() {
+	f := framework.NewDefaultFramework("container-restart")
+	f.NamespacePodSecurityLevel = admissionapi.LevelPrivileged
+
+	ginkgo.Context("Container restart backs off", func() {
+
+		ginkgo.BeforeEach(func() {
+			if err := resetCRIProxyInjector(e2eCriProxy); err != nil {
+				ginkgo.Skip("Skip the test since the CRI Proxy is undefined.")
+			}
+		})
+
+		ginkgo.AfterEach(func() {
+			err := resetCRIProxyInjector(e2eCriProxy)
+			framework.ExpectNoError(err)
+		})
+
+		ginkgo.It("Container restart backs off.", func(ctx context.Context) {
+			// 0s, 0s, 10s, 30s, 70s, 150s, 310s
+			doTest(ctx, f, 5, containerName, 7)
+		})
+	})
+
+	ginkgo.Context("Alternate container restart backs off as expected", func() {
+
+		tempSetCurrentKubeletConfig(f, func(ctx context.Context, initialConfig *kubeletconfig.KubeletConfiguration) {
+			initialConfig.CrashLoopBackOff.MaxContainerRestartPeriod = &metav1.Duration{Duration: time.Duration(30 * time.Second)}
+			initialConfig.FeatureGates = map[string]bool{"KubeletCrashLoopBackOffMax": true}
+		})
+
+		ginkgo.BeforeEach(func() {
+			if err := resetCRIProxyInjector(e2eCriProxy); err != nil {
+				ginkgo.Skip("Skip the test since the CRI Proxy is undefined.")
+			}
+		})
+
+		ginkgo.AfterEach(func() {
+			err := resetCRIProxyInjector(e2eCriProxy)
+			framework.ExpectNoError(err)
+		})
+
+		ginkgo.It("Alternate restart backs off.", func(ctx context.Context) {
+			// 0s, 0s, 10s, 30s, 60s, 90s, 120s, 150, 180, 210)
+			doTest(ctx, f, 7, containerName, 10)
+		})
+	})
+})
+
+func doTest(ctx context.Context, f *framework.Framework, targetRestarts int, containerName string, maxRestarts int) {
+
+	pod := e2epod.NewPodClient(f).Create(ctx, newFailAlwaysPod())
+	podErr := e2epod.WaitForPodContainerToFail(ctx, f.ClientSet, f.Namespace.Name, pod.Name, 0, "CrashLoopBackOff", 1*time.Minute)
+	gomega.Expect(podErr).To(gomega.HaveOccurred())
+
+	// Wait for 150s worth of backoffs to occur so we can confirm the backoff growth.
+	podErr = e2epod.WaitForContainerRestartedNTimes(ctx, f.ClientSet, f.Namespace.Name, pod.Name, "restart", 150*time.Second, targetRestarts)
+	gomega.Expect(podErr).ShouldNot(gomega.HaveOccurred(), "Expected container to repeatedly back off container failures")
+
+	r, err := extractObservedBackoff(ctx, f, pod.Name, containerName)
+	framework.ExpectNoError(err)
+
+	gomega.Expect(r).Should(gomega.BeNumerically("<=", maxRestarts))
+}
+
+func extractObservedBackoff(ctx context.Context, f *framework.Framework, podName string, containerName string) (int32, error) {
+	var r int32
```

Comment:
Where is r set? Am I missing something?

Reply:
Yes this should have been set in L97. I've pushed the update now.

This was actually obfuscated earlier to me because of a DIFFERENT mistake, which is fixed in [2732d57](https://github.com/kubernetes/kubernetes/pull/128559/commits/2732d57e33fd8c45a6f4856fce52f7c8ed8073c9) (I was polling the wrong container name, `restarts` instead of `restart` 😭 )

I do think between this being correct, the mistake in [2732d57](https://github.com/kubernetes/kubernetes/pull/128559/commits/2732d57e33fd8c45a6f4856fce52f7c8ed8073c9) fixed, and the timeout extended to 210s (which better showcases the differences between the test case with the default 300s max and the test case with the configured 30s max, because the time that those two cases become obviously different is only after ~150s), this variation of the test is actually better/more clear than what I suggested in https://github.com/lauralorenz/kubernetes/compare/crashloopbackoff-refactorimagepullbackoff-e2enodecriproxytest...lauralorenz:kubernetes:crashloopbackoff-refactorimagepullbackoff-e2enodecriproxytest-less-flakey

Reply:
I still don't see it being set?

Reply:
Yeah, I'd prefer to just return 0 and get rid of this variable